src/components/advanced-color-picker/advanced-color-picker.test.tsx

```diff
@@ -0,0 +1,309 @@
+  await user.keyboard(keyCode);
+  expect(screen.getByRole("dialog")).toBeVisible();
+});
+
+test.each(["a", "b", "q", "t", "x", "4", "0"])(
+  "when the %s key is pressed, the dialog should not open",
+  async (key) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+    render(<ControlledColorPicker />);
+    screen.getByRole("button", { name: "Change colour" }).focus();
+    await user.keyboard(key);
+    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
+  }
+);
+
+test("the accessible description of the button includes the name of the currently selected color when none is selected", () => {
+  render(<ControlledColorPicker selectedColor="#EBAEDE" />);
+
+  expect(
+    screen.getByRole("button", { name: "Change colour" })
+  ).toHaveAccessibleDescription("Current colour assigned: orchid");
+});
+
+test.each(availableColors)(
+  "the accessible description of the button includes the name of the currently selected color after the user selects $label as a new color",
+  async ({ label }) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+    render(<ControlledColorPicker />);
+    await user.click(screen.getByRole("button", { name: "Change colour" }));
+    await user.click(screen.getByRole("radio", { name: label }));
+    await user.click(screen.getByRole("button", { name: "Close" }));
+
+    expect(
+      screen.getByRole("button", { name: "Change colour" })
+    ).toHaveAccessibleDescription(`Current colour assigned: ${label}`);
+  }
+);
+
+test("the button has the correct background color matching the initially-selected color", () => {
+  render(<ControlledColorPicker selectedColor="#EBAEDE" />);
+
+  expect(screen.getByRole("button", { name: "Change colour" })).toHaveStyle({
+    backgroundColor: "#EBAEDE",
+  });
+});
+
+test.each(availableColors)(
+  "the button has the correct background color after the user selects $label as a new color",
+  async ({ value, label }) => {
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+    render(<ControlledColorPicker />);
+    await user.click(screen.getByRole("button", { name: "Change colour" }));
+    await user.click(screen.getByRole("radio", { name: label }));
+    await user.click(screen.getByRole("button", { name: "Close" }));
+
+    const transparentSvg =
+      "%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%" +
+      "2Fsvg%22%20width%3D%22400%22%20height%3D%22400%22%20fill-opacity%3D%22." +
+      "45%22%3E%3Crect%20x%3D%22200%22%20width%3D%22200%22%20height%3D%22200%22%20%2" +
+      "F%3E%3Crect%20y%3D%22200%22%20width%3D%22200%22%20height%3D%22200%22%20%2F%3E%3C%2Fsvg%3E";
+
+    const styleProperty =
+      value === "transparent" ? "backgroundImage" : "backgroundColor";
+    const styleValue =
+      value === "transparent"
+        ? `#eee url("data:image/svg+xml,${transparentSvg}")`
+        : value;
```

Comment:
thanks for the comment. I agree the conditional logic isn't great, and I thought of doing this myself. The only reason I didn't was that I didn't like the alternative, which was moving the logic to a `.filter` in the `test.each` to remove the `transparent` color from the ones being tested in the non-exceptional case.

Thinking about it again, I guess I can just make a separate const for the array of non-transparent colors, and still keep `availableColors` with all of them as a separate const that uses the other (especially because it's not important that the `transparent` option appears in the middle of that array, so I can just use spread syntax to add it to the start or end). I think this would be a good option, better than both what I currently have and using `filter` in the `test.each` to remove `transparent` - but let me know if you disagree.

Reply:
I agree that defining the non-transparent colors separately would enhance simplicity and clarity.

I suggest removing the `availableColors` array and specifying the colours directly within the tests to improve readability. While this approach may introduce some duplication, the benefits of reducing file-scope variables and defining test parameters within the tests outweigh this minimal duplication. This change would result in more self-contained tests that do not rely on external variables, thereby enhancing overall code readability and maintainability.

Reply:
That's a fair point - and in general I agree that it's worth accepting a small amount of duplication for the same of keeping tests self-contained. I am though a bit reluctant here to redefine the array in every single test, just because it's a relatively large amount of code that's being repeated across the tests, and the precise content isn't even that important to most or all of these test.

I'll give it a go though and decide if I can live with it 👍

Reply:
update: I've now implemented this and have decided I'm happy with the repetition 👍